dali/pipeline/executor/executor2/exec_node_task.cc

```diff
@@ -0,0 +1,507 @@
+      if (inp.event)
+        events.insert(inp.event);
+      ws_->SetOutput(o, inp.data);
+    } else {
+      assert(ws_->OutputIsType<GPUBackend>(o));
+      auto &inp = TaskInput<GPUBackend>(o);
+      if (inp.event)
+        events.insert(inp.event);
+      ws_->SetOutput(o, inp.data);
+    }
+  }
+
+  for (auto e : events)
+    ws_->output_order().wait(e);
+
+  for (int o = 0; o < ws_->NumOutput(); o++) {
+    if (ws_->OutputIsType<CPUBackend>(o)) {
+      auto &out = ws_->Output<CPUBackend>(o);
+      out.set_order(ws_->output_order(), false);
+    } else {
+      auto &out = ws_->Output<GPUBackend>(o);
+      out.set_order(ws_->output_order(), false);
+    }
+  }
+
+  std::optional<int> device = {};
+  if (ws_->output_order().is_device()) {
+    assert(ws_->event());
+    device = ws_->output_order().device_id();
+    CUDA_CALL(cudaEventRecord(ws_->event(), ws_->output_order().stream()));
+  }
+
+  PipelineOutput ret{ *ws_, event_, device };
+  return ret;
+}
+
+//////////////////////////////////////////////////////////////////////////////////
+// ExecNodeTask
+
+tasking::SharedTask ExecNodeTask::CreateTask(ExecNode *node, const WorkspaceParams &params) {
+  if (node->is_pipeline_output) {
+    return tasking::Task::Create(
+      OutputTask(node, params).GetRunnable());
+  } else {
+    int nout = node->outputs.size();
+    return tasking::Task::Create(
+      nout,
+      OpTask(node, params).GetRunnable());
+  }
+}
+
+void ClearWorkspacePayload(Workspace &ws) {
+  auto event = ws.has_event() ? ws.event() : nullptr;
+  for (int i = 0; i < ws.NumInput(); i++) {
+    // TODO(michalz): Some smarter deletion management
+    // If an input has multiple consumers, we should guarantee that the buffer is not destroyed
+    // before all consumers are done with it. A simple way of achieving that is waiting in the
+    // stream associated with the buffer for the consumer's completion event.
+    // A low-hanging optimization is to skip that when the buffer's associated stream is
+    // the same stream as the one the consumer executes on. Doing so will create bubbles in
+    // per-operator streams.
+    // Perhaps a better approach would be to move the inputs to a dedicated deletion stream.
+    // on which we would record the completion events prior to decreasing the reference count.
+    if (ws.InputIsType<CPUBackend>(i)) {
+      if (auto &pinp = ws.InputPtr<CPUBackend>(i)) {
+        auto &inp = *pinp;
+        if (event && inp.order() != ws.output_order())
+          inp.order().wait(event);
+        ws.SetInput<CPUBackend>(i, nullptr);
+      }
+    } else if (ws.InputIsType<GPUBackend>(i)) {
+      if (auto &pinp = ws.InputPtr<GPUBackend>(i)) {
+          auto &inp = *pinp;
+        if (event && inp.order() != ws.output_order())
+          inp.order().wait(event);
+        ws.SetInput<GPUBackend>(i, nullptr);
+      }
+    }
+  }
+
+  for (int i = 0; i < ws.NumArgumentInput(); i++)
+    ws.SetArgumentInput(i, nullptr);
```

Comment:
Here we don't worry about releasing them too early?

Reply:
Good point, I guess? Still, I don't think we issue direct H2D copies of argument inputs or otherwise use (pinned) argument inputs in device code - and that's what would cause problems.